src/creation/creator.rs

```diff
@@ -0,0 +1,301 @@
+use crate::{
+    nodes::{NodeCount, NodeIndex, NodeMap},
+    units::{ControlHash, PreUnit, Unit},
+    Hasher, Round,
+};
+use log::trace;
+
+pub(super) struct Creator<H: Hasher> {
+    node_id: NodeIndex,
+    n_members: NodeCount,
+    candidates_by_round: Vec<NodeMap<Option<H::Hash>>>,
+    n_candidates_by_round: Vec<NodeCount>, // len of this - 1 is the highest round number of all known units
+}
+
+impl<H: Hasher> Creator<H> {
+    pub(super) fn new(node_id: NodeIndex, n_members: NodeCount) -> Self {
+        Creator {
+            node_id,
+            n_members,
+            candidates_by_round: vec![NodeMap::new_with_len(n_members)],
+            n_candidates_by_round: vec![NodeCount(0)],
+        }
+    }
+
+    fn current_round(&self) -> Round {
+        (self.n_candidates_by_round.len() - 1) as Round
+    }
+
+    // initializes the vectors corresponding to the given round (and all between if not there)
+    fn init_round(&mut self, round: Round) {
+        if round > self.current_round() {
+            let new_size = (round + 1).into();
+            self.candidates_by_round
+                .resize(new_size, NodeMap::new_with_len(self.n_members));
+            self.n_candidates_by_round.resize(new_size, NodeCount(0));
+        }
+    }
+
+    pub(super) fn create_unit(&self, round: Round) -> (PreUnit<H>, Vec<H::Hash>) {
+        let parents = {
+            if round == 0 {
+                NodeMap::new_with_len(self.n_members)
+            } else {
+                self.candidates_by_round[(round - 1) as usize].clone()
```

Comment:
Thanks for the comments, very useful.
* If we export it as `pub` we would have to document & support it. Also, I'm pretty sure some lint will be unhappy about leaking private types in public interfaces.
* Fixed, now it checks whether it `can_create` before actually doing anything, if it cannot it returns `Err(())`.
* This is mostly by design, I didn't want `Creator` to contain knowledge about the round, since this is being controlled outside. I can redesign this, with the main tradeoff (I think?) being making the creation of forks harder vs making creator itself more complicated. I would prefer for it to stay as-is, but I'm open to arguments to the contrary.
* Removed `is_behind`, made `current_round` public. Now `run` checks the condition explicitly. I think this is actually better.
* This assumption is wrong afaik, you can receive an ancient unit after a long time and then the units will not be passed round-by-round.
Could you make comments in separate, uh, comments in the future? It's easier to keep the discussion organized then.

Reply:
`pub` method of a non-`pub`/non-`exported` type is not `pub` externally?

Reply:
> This assumption is wrong afaik, you can receive an ancient unit after a long time and then the units will not be passed round-by-round.
Let me reword this: round-by-round, i.e. when you receive a unit from round `r+1` in the creator, then you know that you're able to create a unit for round `r` (but you can still wait for some more unit if you want). My intention was also: if it would be redesigned this way, then it would be not-possible to create forks (data for a round in which we already created a unit is disposed).

Reply:
> pub method of a non-pub/non-exported type or module is not pub externally?

Clippy still complains about leakage. It's also potentially less readable, especially if we do it differently in different modules?

I think this would work with clippy if we rewrote everything to the setup where everything is pub within private modules and then only the things we want exported are reexported in the module. This would require quite a lot of work though, and I'm not even sure it would be better. Some standards would be nice, as usual.

> Let me reword this:

Ah, I see what you mean. Yeah, we could do that although currently tests don't keep that assumption. Also, I think you are making a broader point about how you imagine the design/API of creator that I am not quite getting – if that is right, lets discuss this in a more live setting.